### PR TITLE
Kemp/gradient optimization

### PR DESCRIPTION
**Pull request by tdkemp**
@Besler we have an unconstrained gradient descent optimizer working for the constant propagation model. It is not very compact, so I'd appreciate your feedback on ways to make it better.

1) It seems I have to re-call the LS solver in every iteration of the while loop, so it adds a lot of redundant code. If I don't do this, though, the update does not work.
2) Is there any way to generalize the model we use? Instead of having a specific optimizer for the constant propagation model and another optimizer for a curvature model, can we just have one optimizer and tell it which model we're working with? I couldn't quite wrap my head around this because we're updating the constant in the LS solver every iteration.
3) It is very slow because we need to run the LS solver twice per iteration, I don't think there's any way around it.
4) I did some basic checks and it seems to be convex which is great!
5) The optimizer is currently unconstrained so I still need to add in the bounds, I'd like to discuss this with you as I have some ideas.

### CODE REVIEW

File: wolff/solver.py [hunk 1]
```diff
@@ -6,14 +6,19 @@
 
 
 class Solver(object):
-    def __init__(self, simulation_time=0.0, reinitialization_iters=0):
+    def __init__(self,
+                 simulation_time=0.0,
+                 reinitialization_iters=0,
+                 target_times=[]
```
Comment (Besler) [inline wolff/solver.py]: Okay, I think we should change this slightly to only have `simulation_time` (and rename it `simulation_times`) or just have `reinitialization_iters`. I prefer `simulation_times`. It's backwards incompatible so we have to go chase down all occurrences of `simulation_time`, but that's okay.

File: wolff/solver.py [hunk 2]
```diff
@@ -113,6 +133,14 @@ def simulation_time(self, value):
...
         self._simulation_time = value
 
+    @property
+    def target_times(self):
+        return self._target_times
+
+    @target_times.setter
+    def target_times(self, value):
+        self._target_times = value
+
```
Comment (Besler) [inline wolff/solver.py]: If we do change to 'simulation_times`, we can change the setter such that if we get a scalar, we force it into a list or if we get  a list, we just assign it. Must copy the list because they are mutable in Python.
```suggestion
    @property
    def target_times(self):
        return self._target_times

    @target_times.setter
    def target_times(self, value):
        if isinstance(value, list):
            # Deep copy list
            self._target_times = list(value)
        else:
            # Assume constant number
            self._target_times =[value]

```

File: wolff/solver.py [hunk 3]
```diff
@@ -73,6 +78,21 @@ def halt(self):
...
+                    and (self._elapsed_time < self.target_times[i+1]):
+                break
+
+        dt = np.amin(np.array([
+            self.model.dt,
+            self.simulation_time - self._elapsed_time,
+            self.target_times[i+1] - self._elapsed_time])
+        )
+
+        return dt
```
Comment (Besler) [inline wolff/solver.py]: Great solution. This will simplify by having everything in one list.

File: wolff/optimization/yang_cost_function.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+from wolff.layer.layer import Layer
+from wolff.layer.heaviside_step_function import HeavisideStepFunction
+import numpy as np
+
+
+class YangCostFunction(Layer):
+    def __init__(self, grids_predicted=[],
+                 grids_true=[],
+                 elapsed_times_predicted=[],
+                 elapsed_times_true=[]):
```
Comment (Besler) [inline wolff/optimization/yang_cost_function.py]: Correct me if I'm wrong, but for the Yang _et al_ cost function, the true and predicted elapsed times have to be equivalent, no? Can we just have one set of times here?

File: wolff/optimization/gradient_descent.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        self.grad_J = grad_J
+        self.learning_rate = learning_rate
+
+    @property
+    def theta(self):
+        return self._theta
+
+    @theta.setter
+    def theta(self, value):
+        self._theta = value
```
Comment (Besler) [inline wolff/optimization/gradient_descent.py]: Since this is a list, we need a deep copy.
```suggestion
        self._theta = list(value)
```

File: wolff/optimization/gradient_descent.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    @property
+    def grad_J(self):
+        return self._grad_J
+
+    @grad_J.setter
+    def grad_J(self, value):
+        self._grad_J = value
+
+    def process(self):
+        if self.theta is None:
```
Comment (Besler) [inline wolff/optimization/gradient_descent.py]: These will never be `None` since their default value is `[]`

File: wolff/optimization/gradient_descent.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            raise ValueError('Constant must be greater than zero')
+        self._learning_rate = value
+
+    @property
+    def grad_J(self):
+        return self._grad_J
+
+    @grad_J.setter
+    def grad_J(self, value):
+        self._grad_J = value
```
Comment (Besler) [inline wolff/optimization/gradient_descent.py]: Since this is a list, we need a deep copy.
```suggestion
        self._grad_J = list(value)
```

File: wolff/optimization/finite_difference_gradient.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            raise ValueError('dtheta must be greater than zero')
+        self._dtheta = value
+
+    @property
+    def J(self):
+        return self._J
+
+    @J.setter
+    def J(self, value):
+        self._J = value
```
Comment (Besler) [inline wolff/optimization/finite_difference_gradient.py]: Since this is a list, we need a deep copy.
```suggestion
        self._J = list(value)
```

File: wolff/optimization/finite_difference_gradient.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    @property
+    def J_dtheta(self):
+        return self._J_dtheta
+
+    @J_dtheta.setter
+    def J_dtheta(self, value):
+        self._J_dtheta = value
```
Comment (Besler) [inline wolff/optimization/finite_difference_gradient.py]: Since this is a list, we need a deep copy.
```suggestion
        self._J_dtheta = list(value)
```

File: wolff/optimization/constant_propagation_finite_difference_solver.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+    import PropagationModel
+from wolff.sinks.grid_sink import GridSink
+from wolff.optimization.yang_cost_function import YangCostFunction
+from wolff.reinitializers.peng_reinitialization import PengReinitialization
+from wolff.optimization.gradient_descent import GradientDescent
+from wolff.optimization.finite_difference_gradient \
+    import FiniteDifferenceGradient
+
+
+class ConstantPropagationFiniteDifferenceSolver(Layer):
```
Comment (Besler) [inline wolff/optimization/constant_propagation_finite_difference_solver.py]: Is there a reason you cannot inherit from `Solver` and avoid a lot of this repeated code?

Comment (Besler) [reply wolff/optimization/constant_propagation_finite_difference_solver.py]: Also, I think it would be easier to make this simpler. We can talk abstraction.

File: examples/constant_propagation_optimization.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+F.dtheta = dtheta
+F.learning_rate = learning_rate
+F.initial_speed = initial_speed
+F.max_iterations = max_iterations
+F.epsilon = epsilon
+F.grids_true = grids_true
+F.elapsed_times_true = elapsed_times_true
+
+print('Optimizing')
+predicted_speed = F()
```
Comment (Besler) [inline examples/constant_propagation_optimization.py]: Print predicted speed at the end?

File: wolff/optimization/yang_cost_function.py (hunk 4 above)
Comment (tdkemp) [reply wolff/optimization/yang_cost_function.py]: We need both, so we can solve the model for any given amount of experimental data.

**Comment by tdkemp** [comment]
@Besler I made some of the changes we talked about last week, so I'd like to hear your feedback on the updates. I'm still working on an adaptive learning rate/ adding momentum.

p.s. curvature + propagation optimizer is working.

File: wolff/optimization/constrained_gradient_descent_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    def theta_0(self, value):
+        self._theta_0 = value
+
+    @property
+    def grids_true(self):
+        return self._grids_true
+
+    @grids_true.setter
+    def grids_true(self, value):
+        self._grids_true = value
```
Comment (Besler) [inline wolff/optimization/constrained_gradient_descent_optimizer.py]: Should we copy all data or not here?

File: wolff/optimization/constrained_gradient_descent_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+    @property
+    def elapsed_times_true(self):
+        return self._elapsed_times_true
+
+    @elapsed_times_true.setter
+    def elapsed_times_true(self, value):
+        self._elapsed_times_true = value
```
Comment (Besler) [inline wolff/optimization/constrained_gradient_descent_optimizer.py]: Should we deep copy the list?
```suggestion
        self._elapsed_times_true = list(value)
```

File: wolff/optimization/curvature_finite_difference_gradient_model.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    def grids_true(self, value):
+        self._grids_true = value
+
+    @property
+    def elapsed_times_true(self):
+        return self._elapsed_times_true
+
+    @elapsed_times_true.setter
+    def elapsed_times_true(self, value):
+        self._elapsed_times_true = value
```
Comment (Besler) [inline wolff/optimization/curvature_finite_difference_gradient_model.py]: Same as before about copying lists?
```suggestion
        self._elapsed_times_true = list(value)
```

File: wolff/optimization/curvature_finite_difference_gradient_model.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    def theta(self, value):
+        self._theta = list(value)
+
+    @property
+    def grids_true(self):
+        return self._grids_true
+
+    @grids_true.setter
+    def grids_true(self, value):
+        self._grids_true = value
```
Comment (Besler) [inline wolff/optimization/curvature_finite_difference_gradient_model.py]: Do we want a deep copy here?

File: wolff/solver.py [hunk 16]
```diff
@@ -7,18 +7,17 @@
...
     def __init__(self,
-                 simulation_time=0.0,
+                 simulation_times=[0.0],
                  reinitialization_iters=0,
                  target_times=[]
                  ):
         # Gettable variables
-        self.simulation_time = simulation_time
+        self.simulation_times = simulation_times
         self._model = None
```
Comment (Besler) [inline wolff/solver.py]: Did you change all instances of `solver.simulation_time` to `solver.simulation_times` in all examples and the command line interfaces?

File: wolff/optimization/constrained_gradient_descent_optimizer.py (hunk 12 above)
Comment (tdkemp) [reply wolff/optimization/constrained_gradient_descent_optimizer.py]: Yes, thanks for noticing those!

File: wolff/solver.py (hunk 16 above)
Comment (tdkemp) [reply wolff/solver.py:15]: Sure, I made those updates.

**Comment by tdkemp** [comment]
@Besler I actually made a few small changes this morning. It's now an Adam solver, so much faster than SDG.

**Comment by Besler** [comment]
Really impressed that you implemented your own Adam optmizer. I hate to see you throw away all the code you wrote. If there is a way to abstract the "optimizer" part, that might be worthwhile.

For instance, you could define a generic optimizer class that does everything but compute the update.
```python
@abstractmethod
def next_theta(self, theta, gradient_vector):
    # Return dTheta
    # Simple case: return theta + \eta gradient_vector
```
Then you can have an "Adam optimizer" and "SGD" optimizer inherit and just implement this method.

You'll find a nice way to do it, if you care to have it.

**Comment by tdkemp** [comment]
Great idea, I'll take a look at that and see what I can put together.

**Comment by tdkemp** [comment]
I abstracted the gradient descent optimizer and added an Adam, momentum and SDG optimizer that inherit from the base optimizer. If you approve, I'll merge.

File: examples/constant_propagation_optimization.py [hunk 17]
```diff
@@ -70,8 +70,7 @@
 S.grids_true = grids_true
 S.elapsed_times_true = elapsed_times_true
 S.grid = DistanceTransform()(rod.Copy())
+S.print_output = True
 
 print('Optimizing')
 predicted_speed = S.simulate()
-
-print(predicted_speed)
```
Comment (Besler) [inline examples/constant_propagation_optimization.py]: Removed print?

File: wolff/optimization/gradient_descent_solver.py [hunk 18]
```diff
@@ -136,31 +154,44 @@ def simulate(self):
...
+
+            self.Print()
 
     def halt(self):
         return (self._error < self._epsilon) or (self._iters > self._max_iters)
+
+    def Print(self):
+        if self._print_output:
+            print('Iteration {}: error = {}\t theta = {}'.
+                  format(self._iters, self._error, self._theta))
```
Comment (Besler) [inline wolff/optimization/gradient_descent_solver.py:190]: Very nice

File: wolff/optimization/momentum_optimizer.py [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+        if value <= 0:
+            raise ValueError('Momentum rate must be greater than zero')
+        self._momentum_rate = value
+
+    def next_theta(self):
+
+        current_m = self._model.H.data + \
+            self._momentum_rate * self._constants[0]
+
+        self._theta = self._theta - (self._learning_rate * current_m)
```
Comment (Besler) [inline wolff/optimization/momentum_optimizer.py:40]: Are you confident in this math? I seem to recall something more like this, but I'm only glancing.

Momentum

File: wolff/optimization/sdg_optimizer.py [hunk 20]
```diff
@@ -0,0 +1,10 @@
+from wolff.optimization.gradient_descent_solver \
+    import GradientDescentSolver
+
+
+class SDGOptimizer(GradientDescentSolver):
```
Comment (Besler) [inline wolff/optimization/sdg_optimizer.py:5]: `SDG` or `SGD`?


(sorry)

File: examples/constant_propagation_optimization.py (hunk 17 above)
Comment (tdkemp) [reply examples/constant_propagation_optimization.py]: Added it back.

File: wolff/optimization/sdg_optimizer.py (hunk 20 above)
Comment (tdkemp) [reply wolff/optimization/sdg_optimizer.py:5]: Yup, should be fixed now.

File: wolff/optimization/momentum_optimizer.py (hunk 19 above)
Comment (tdkemp) [reply wolff/optimization/momentum_optimizer.py:40]: Also my mistake, thanks for catching that.

**Comment by Besler** [comment]
All looks good to me! Merge when you're happy.
